### PR TITLE
Django18 upgrade

### PR DESCRIPTION
**Pull request by aleksandra-tarkowska**
This PR helps choosing right Django version depends on Python.
 - Python 2.6 Django 1.6
 - Python 2.7 Django 1.8

The aim of this PR is:
 - to deploy Omero.web using both python versions,
 - clarify use of ``request.GET`` and ``request.POST``
 - fix deprecated warnings.

To test:
- check if https://ci.openmicroscopy.org/view/DEV/job/OMERO-DEV-merge-daily/
  - OMERO-DEV-merge-integration-Python27 -> https://cowfish/integration
  - OMERO-DEV-merge-integration - > https://trout/integration
  are green
- log in to both Omero.web

To test web plugin:
- uninstall django
- bin/omero web start should tells ``Django is not installed!``.

-----------

**Known issues:**
 - request.REQUEST still in webgateway_cache. That will be fixed in another PR if ok

---------------

**To test ui:**

**testing servers: Django 1.6 on eel/merge and Django 1.8 on cowfish/integration
when reporting bugs please indicate which server was used and exact timestamp when error occurred**

- feedback: send comment and error
 - go to https://eel and cowfish/merge or integration/feedback/comment/ and send comment.
 - go to https://eel and cowfish/merge or integration/weberror/error_500/ and submit error.
- login:
 - go to https://eel and cowfish/merge or integration/webclient/login/?url=%2Fmerge%2Fwebclient%2Fuserdata%2F%3Fexperimenter%3D-1, log in, you should be redirected to https://eel and cowfish/merge or integration/webclient/userdata/?experimenter=-1
- webadmin:
 - Reset Password https://www.openmicroscopy.org/private/ome-internal/testing_scenarios/Resetpwd.html
 - add new user / group https://www.openmicroscopy.org/private/ome-internal/testing_scenarios/AdminGroupUserCreate.html (handled by robot test)
 - add new user / group https://www.openmicroscopy.org/private/ome-internal/testing_scenarios/AdminGroupUserEdit.html (handled by robot test)
 - edit my account (handled by robot test)
 - crop avatar
- webclient:
 - check direct link takes to the image webclient/userdata/?show=image-123
 - search using global searching as well as advance searching
 - annotation (test with multiple selected images) https://www.openmicroscopy.org/private/ome-internal/testing_scenarios/Annotate.html
 - create Project, Dataset, Screen
 - create and edit Share https://www.openmicroscopy.org/private/ome-internal/testing_scenarios/Share.html
 - edit channel name
 - rendering settings: copy, paste, save, apply to all 
 - download:
   - attachement
    - original file
    - export to
- scripts https://www.openmicroscopy.org/private/ome-internal/testing_scenarios/ScriptsClient.html

### CODE REVIEW

**Comment by manics** [comment]
RFE for a future PR: It looks like there's a duplication of logic in finding the paths- move it to a common method?

**Comment by aleksandra-tarkowska** [comment]
omero_ext or omero.utils?

**Comment by manics** [comment]
Maybe `utils`? I'm not bothered as far as this PR is concerned (though others might be), it just struck me that `omero` `cli` and `setup.py` and all combining paths for `lib`, `python`, `fallback`, etc and should probably be refactored at some point.

**Comment by sbesson** [comment]
Discussing with @aleksandra-tarkowska @manics @manics @jburel, next step is to work on migrating Django as a traditional Python dependency which needs to be installed by the sysadmin (pip/virtualenv/conda...) instead of bundling it with our application. /cc @chris-allan @knabar

**Comment by aleksandra-tarkowska** [comment]
how about if we add ``bin/omero install`` that would call ``pip install --install-option="--prefix=lib/python" -r req.txt``

**Comment by manics** [comment]
That'll only work if it's already running in a virtualenv, you won't be able to use the system pip

**Comment by aleksandra-tarkowska** [comment]
Django version is now available in diagnostics

**Comment by aleksandra-tarkowska** [comment]
These are all the changes required for django upgrade.

**Comment by sbesson** [comment]
Note https://ci.openmicroscopy.org/job/OMERO-VIRTUALBOX-build/327/ is now red because Django is missing in the base image. Should the requirements be specified for each supported plaform in https://github.com/ome/ome-veewee @manics or just Ubuntu ?

**Comment by manics** [comment]
Maybe we should have the vbox vs docker discussion sooner rather than later.

**Comment by will-moore** [comment]
Looks good.

**Comment by sbesson** [comment]
Assuming everything is green tonight, what are the next steps here in terms of testing?

**Comment by aleksandra-tarkowska** [comment]
I'd move it out of breaking. Then go through the list of tests in the description

**Comment by aleksandra-tarkowska** [comment]
We will need integration and robot running agains both pythons

**Comment by sbesson** [comment]
:+1: let's discuss post-meeting how to reorganize the Ci infrastructure.

**Comment by sbesson** [comment]
Discussed with @aleksandra-tarkowska, moving this PR out of breaking for review in the traditional merge build. Goal is to add set up another integration stack for Python 2.7.

**Comment by sbesson** [comment]
Given the state of https://ci.openmicroscopy.org/job/OMERO-DEV-breaking-trigger/498/, moving out of `breaking`.

File: components/tools/OmeroWeb/setup.py [hunk 1]
```diff
@@ -17,17 +17,22 @@
     if tools.find(".".join(map(str, sys.version_info[0:2]))) > 0:
         sys.path.insert(0, os.path.abspath(tools))
 
-os.environ.setdefault('OMERO_HOME', os.path.abspath(
-    os.path.join("..", "..", "..", "dist")))
+if "test" in sys.argv:
```
Comment (aleksandra-tarkowska) [inline components/tools/OmeroWeb/setup.py]: This is required because this file is loaded on tools-clean

File: components/tools/OmeroWeb/setup.py [hunk 2]
```diff
@@ -17,17 +17,22 @@
...
+    sys.path.insert(0, LIB)
+    OMEROWEB_LIB = os.path.join(LIB, "omeroweb")
+    sys.path.insert(1, OMEROWEB_LIB)
 
-os.environ.setdefault("DJANGO_SETTINGS_MODULE", "omeroweb.settings")
+    os.environ.setdefault("DJANGO_SETTINGS_MODULE", "omeroweb.settings")
+
+    import django
+    if django.VERSION > (1,7):
+        django.setup()
```
Comment (aleksandra-tarkowska) [inline components/tools/OmeroWeb/setup.py]: https://docs.djangoproject.com/en/1.8/releases/1.7/#standalone-scripts

**Comment by aleksandra-tarkowska** [comment]
test_chgrp_* fixed 
```
test/integration/test_chgrp.py ..g

============================================= 5 tests deselected by '-ktest_chgrp_new_container' =============================================
=================================================== 2 passed, 5 deselected in 3.67 seconds ===================================================
```

**Comment by aleksandra-tarkowska** [comment]
```
$ git checkout django18_upgrade
Switched to branch 'django18_upgrade'
omeroweb $  git grep -E "request.REQUEST"
```
give nothing

**Comment by aleksandra-tarkowska** [comment]
With the above changes and few extra fixes Robot tests are passing locally

**Comment by aleksandra-tarkowska** [comment]
Last commit should fix https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-integration-web/96/

**Comment by aleksandra-tarkowska** [comment]
Robot is green https://ci.openmicroscopy.org/view/DEV/job/OMERO-DEV-merge-robotframework/116/robot/report/log.html

**Comment by pwalczysko** [comment]
Test of ``feedback`` (UI testing in the header of this PR) passed

**Comment by pwalczysko** [comment]
Test of ``login`` (UI testing in ;the header of this PR) passed.

**Comment by pwalczysko** [comment]
Testing ``Reset password`` as per https://www.openmicroscopy.org/private/ome-internal/testing_scenarios/Resetpwd.html. After I have put in a valid username and valid email address (which I have registered with, I get

screen shot 2015-10-12 at 15 14 32

@aleksandra-tarkowska Expected ?

**Comment by pwalczysko** [comment]
Cannot repeat https://github.com/openmicroscopy/openmicroscopy/pull/4236#issuecomment-147410761

**Comment by pwalczysko** [comment]
Deleted all previous comments. They were not relevant, as wrong server was used.

**Comment by pwalczysko** [comment]
Workflow to repeat tomorrow just after standup before the testing starts.

   - on eel merge, login as user-6
   - create a  user (e.g. ``test-ola3``) and add it to ``private-1`` group
   - set a valid email address to this user (petr's email for example)
   - logout 
   - do not login as anyone, but click ``forgotten password`` link
   - put in the ``test-ola3`` and the valid email used credentials into the box
   - watch for a ``Error, contac admin...`` message in the UI and logs

**Comment by aleksandra-tarkowska** [comment]
@pwalczysko successfully reseted password for test user
firefoxscreensnapz009
```
Dear Test Ola3 (test-ola3) your new password is: xxx
```

**Comment by joshmoore** [comment]
Minor comment: perhaps rename the Py27 requirements file to just `requirements.txt` for supporting us moving forward?

**Comment by pwalczysko** [comment]
The testing sheet is https://docs.google.com/spreadsheets/d/1_tkRPZ-EhnoD020EDgwVAwuLf-CPF5Z0UX7Nb9ayIg0/edit#gid=0

**Comment by pwalczysko** [comment]
Problem reported in Column B of https://docs.google.com/spreadsheets/d/1_tkRPZ-EhnoD020EDgwVAwuLf-CPF5Z0UX7Nb9ayIg0/edit#gid=385968527 : wrong form, just pasting the screenshot here as well for better ease of finiding it here.

wrong position of password in the form

**Comment by aleksandra-tarkowska** [comment]
@pwalczysko in terms of annotations, please just follow regular rating, tagging, comment, attachment scenarios on ONE image and with multiple selected images (that is what I meant bulk annotation). This PR is not compatible with 5.1 line and cannot be rebased on metadata branch.
cc: @will-moore @joshmoore

**Comment by pwalczysko** [comment]
Testing is finished. See what was tested in https://docs.google.com/spreadsheets/d/1_tkRPZ-EhnoD020EDgwVAwuLf-CPF5Z0UX7Nb9ayIg0/edit#gid=0. 
See the two problems reported in https://docs.google.com/spreadsheets/d/1_tkRPZ-EhnoD020EDgwVAwuLf-CPF5Z0UX7Nb9ayIg0/edit#gid=385968527.
   - wrong form (see screenshot in https://github.com/openmicroscopy/openmicroscopy/pull/4236#issuecomment-147668137) and column B in the sheet
   - ``Save to All`` problem reported by @bramalingam in column C

Note that each of these two problems is specific for one particular server only (one cowfish, one eel) - see sheet for details.

**Comment by aleksandra-tarkowska** [comment]
As discussed with @joshmoore @jburel and @pwalczysko, it is preferred to merge PR in asap and outstanding issues listed during testing should be fixed in a separate PR.

**Comment by aleksandra-tarkowska** [comment]
This PR become a blocker for https://github.com/openmicroscopy/openmicroscopy/pull/4259 because of requirement.txt

**Comment by pwalczysko** [comment]
Created https://trac.openmicroscopy.org/ome/ticket/13063 and https://trac.openmicroscopy.org/ome/ticket/13062. From my point of view (remembering that one of the tickets pasted here is a blocker-regression to be fixed on Monday), ready to merge.

**Comment by aleksandra-tarkowska** [comment]
Thank you everyone for help. More is coming ;-)

**Pull request merged**